### PR TITLE
Redesign and simplify Activity Center app db

### LINKED ISSUE #15215: Activity Center app state is unnecessarily complex
# Problem

A couple months ago, the Activity Center app db state was designed to support a different set of features/expectations. Things changed and the AC is now relatively stable in some areas.

The current app db for the AC is more or less like this (I omited some keys):

```clojure
{:activity-center
 {:filter        {:status :unread
                  :type   types/contact-request}
  :notifications {types/no-type
                  {:all    {:cursor   ""
                            :loading? true
                            :data     [...]}
                   :unread {:cursor   ""
                            :loading? false
                            :data     [...]}}

                  types/contact-request
                  {:all    {:cursor   ""
                            :loading? false
                            :data     [...]}
                   :unread {:cursor ""
                            :data   [...]}}

                  types/mention
                  {:all    {:cursor   ""
                            :loading? false
                            :data     [...]}
                   :unread {:cursor   ""
                            :loading? false
                            :data     [...]}}

                  ;; and so on for other types (AC tabs in the UI)
                  }}}
```

As you can see, the UI is tracking the state of every AC tab, along with its loading state and cursor. This means that whenever notifications need to be *reconciled*, **complex logic needs to take place to correctly update the app db**. Over the last couple months this proved to be too complicated and bug-driven, requiring complicated test cases.

On top of that, **it's hard to make the reconciliation performant** without resorting to complementary indexes, otherwise full linear scans need to be performed in lots of collections.

The existing solution also duplicates a lot of notifications (the data is not normalized) and **more memory is consumed**.

### Solution

We can remove a lot of the complexity if we assume that only the **current tab state** need to be kept in memory. When switching tabs, we always re-fetch the data. Given the communication between the backend and frontend is local and rather fast, this gives a good UX.

Proposal:

```clojure
{:activity-center
 {:filter        {:status :unread
                  :type   types/contact-request}
  :cursor        "..."
  :loading?      true
  :notifications [...]}}
```

### PR DESCRIPTION
**Pull request by ilmotta**
Fixes #15215 

### Summary

Redesign the app db state for the Activity Center. Please, see the issue being fixed for more details about the problems being solved.

TL;DR: There's a lot less state to keep track and reconcile and way less nesting in the app db because we're only managing the state of the *current tab*, not all tabs.

Additionally:

- [x] While updating unit tests, found a bug on the sorting notifications' logic.
- [x] While updating unit tests, found a bug where notifications that are not of the type *contact request* were being removed from the app db.
- [x] Fixed regression where pressing on a notification would not open the chat.
- [x] Hardened unit tests.

#### Platforms

- Android
- iOS

##### Non-functional

- Less memory consumption.
- Faster reconciliation of notification coming from signals and from synchronous responses from RPC calls.

### Steps to test

In theory, nothing should be impacted because it's a refactoring. Although I had to carefully re-test everything manually and update unit tests. So I'll kindly request a review from QAs.

status: ready

### CODE REVIEW

File: src/status_im/data_store/activities.cljs
Comment (flexsurfer) [inline src/status_im/data_store/activities.cljs]: type and author could be destructured from the parameter

File: src/status_im2/subs/activity_center_test.cljs
Comment (flexsurfer) [inline src/status_im2/subs/activity_center_test.cljs:68]: yes, that's how we should deal with tests :)

File: src/status_im2/contexts/activity_center/events.cljs
Comment (flexsurfer) [inline src/status_im2/contexts/activity_center/events.cljs]: could we sort in opposite direction? so reverse is not needed

**Comment by flexsurfer** [comment]
i've also noticed, that when we press the bell button, there is a visible delay, like a few second before openning AC screen, and also in re-frisk i saw AC's event which dispatched like 10 times

**Comment by ilmotta** [comment]
> i've also noticed, that when we press the bell button, there is a visible delay, like a few second before openning AC screen, and also in re-frisk i saw AC's event which dispatched like 10 times

Yes and yes for both things @flexsurfer and both problems will be solved eventually (or soon).

The button delay I've solved in a branch that's not ready to become a PR yet. The only way to eliminate the delay was to force re-frame to render the notifications outside the batch that renders the AC bottom sheet. This delay always existed. I hope to get this improvement merged in the next 1 month hopefully.

About the event being dispatched a lot of times. This is the event that gets the counter of notifications for each tab. It was a temporary workaround implemented a couple weeks ago due to the need to display unread indicators on top of each tab (each type of notification). This week or early next week an improved status-go endpoint will be available so clients can request counters for each type of notification in a single call. I'm coordinating this with the desktop team.

Edit: The N RPC calls to fetch counters have an issue I created some time ago https://github.com/status-im/status-mobile/issues/15119

File: src/status_im2/contexts/activity_center/events.cljs
Comment (ilmotta) [reply src/status_im2/contexts/activity_center/events.cljs]: The original implementation did exactly that, but it had a bug caught by the tests. I attempted to quickly understand why and gave up for the moment, but I can take another look.

In theory this is not gonna be a big deal most of the time, because notifications are going to be re-fetched on tab switches and when opening the AC. But, if the user has hundreds of notifications and scroll to the bottom, then when a new one arrives the reconciliation will be a bit slower, but I suspect this is not gonna be the case for most user flows.

IMO, the ideal solution involves NOT storing in the app db all the notifications as the user scrolls and new pages arrive. We should have a cap on the max number of notifications to store in the app db, e.g. using some kind of circular list data structure and dropping the "old" ones as the user scrolls. I'm interested in taking a look at this because it's a problem that I've optimized in the past, but for the web platform.

File: src/status_im2/subs/activity_center_test.cljs
Comment (ilmotta) [reply src/status_im2/subs/activity_center_test.cljs:68]: Oh man, you have no idea how many times the unit tests for event handlers saved me from introducing bugs. But yeah, this one can go away with the simplified app db.

File: src/status_im2/contexts/activity_center/events.cljs
Comment (ilmotta) [reply src/status_im2/contexts/activity_center/events.cljs]: For instance, without a max limit of notifications to store in the app db, if the user leaves the screen turned on in the AC and new signals arrive constantly (imagine you're famous and receives a lot of notifications) the app will eventually start to perform terribly and consume more battery doing all the synchronization logic.

File: src/status_im2/contexts/activity_center/events.cljs
Comment (smohamedjavid) [inline src/status_im2/contexts/activity_center/events.cljs]: `deleted` and `read` can be destructured here

**Review by smohamedjavid — APPROVED** [review]

File: src/status_im2/contexts/activity_center/events.cljs
Comment (ilmotta) [reply src/status_im2/contexts/activity_center/events.cljs]: For the record: How do we reverse a comparison using `juxt` as the key fn? It wasn't obvious to me. `sort-by` uses the `compare` function to compare the value returned by `(juxt :timestamp :id)`. For example, it would compare `[100 "0x1"]` and `[100 "0x2"]` and return `-1`, because "0x1" is *less than* "0x2". To reverse, we need to pass a comparator function that swaps the arguments passed to `compare`.

Sometimes we just forget about old tricks ;)

```clojure
;; Correctly reverses order
(let [coll [{:id "0x1" :timestamp 1}
            {:id "0x2" :timestamp 50}
            {:id "0x3" :timestamp 50}
            {:id "0x4" :timestamp 100}]]
  (sort-by (juxt :timestamp :id)
           #(compare %2 %1)
           coll))
;; => [{:id "0x4", :timestamp 100}
;;     {:id "0x3", :timestamp 50}
;;     {:id "0x2", :timestamp 50}
;;     {:id "0x1", :timestamp 1}]

```

Comment (ilmotta) [reply src/status_im2/contexts/activity_center/events.cljs]: Fixed https://github.com/status-im/status-mobile/pull/15216/commits/394184eff49519eb73248ca75835a38dabf8d58a

**Review by flexsurfer — APPROVED** [review]

**Comment by ilmotta** [comment]
@status-im/mobile-qa, this PR refactored a bunch of things behind the scenes, but as the word *refactor* suggests, the behavior of the app is still the same. I did a fair share of tests and I don't expect any regressions, but we never know right?

If one of you could just run some manual checks in anything related to the AC it would be very helpful to make sure I'm not breaking anything important.

This PR actually fixes two existing problems in `develop`.

1. Fixed regression where pressing on a notification would not open the chat.
2. When notifications had the same timestamp, sorting was performed incorrectly. It's pretty much impossible to reproduce with e2e, so that's why it wasn't caught before.

**Comment by pavloburykh** [comment]
@ilmotta thanx for the PR and for regression fixes! 

Please, take a look at the issue that is related to this fix

> 1. Fixed regression where pressing on a notification would not open the chat.

### ISSUE 1 Chat screen is not loaded after redirection upon tapping accepted CR notif. "Add Unknow to send a message" is displayed instead of active input field. 

1. User A sends CR to User B
2. User B opens AC and accepts CR
3. User B taps on accepted CR notification
4. Observe chat screen is loaded and User B is able to start messaging with User A

**Expected result:** User B is redirected to 1-1 chat with User A. Chat screen is loaded, input field is active, User B is able to start messaging with User A

**Actual result:** User B is redirected to 1-1 chat but chat screen is not loaded (skeleton is displayed). "Add Unknow to send a message" label is displayed instead of active input field. 

[Status-debug-logs (14).zip](https://github.com/status-im/status-mobile/files/10871828/Status-debug-logs.14.zip)

photo_2023-03-02 15 32 13

https://user-images.githubusercontent.com/97245802/222443049-8cf63df0-06ba-4f47-9902-8883dd969c8d.mp4

**Comment by ilmotta** [comment]
Hi @pavloburykh. Just a couple minutes ago I merged the fix for the issue demonstrated in the video you shared.

I'm sorry for the trouble, but I was trying to reduce the delay to open the AC in another PR. As you probably noticed in Android, whenever you press on the bell icon, there's a significant delay to open the AC. There's a hardcoded delay of 250ms, but reducing it to 50ms caused the bug you reported.

>ISSUE 1 Chat screen is not loaded after redirection upon tapping accepted CR notif. "Add Unknow to send a message" is displayed instead of active input field.

I'll investigate this issue now. Thanks

**Comment by pavloburykh** [comment]
> Hi @pavloburykh. Just a couple minutes ago I merged the fix for the issue demonstrated in the video you shared.

@ilmotta thank you for quick fix! I have faced another issue and trying to figure out the steps. could check the logs and tell if it is related to the same issue with reducing the delay?

### ISSUE 2 Delete/Read/Unread swipe notification buttons stop responding at some moment in AC

Still trying to figure out the steps. Basically I have generated about 10+ notifications including different types: CR notifs, admin, mentions, replies. A was interacting with those notifications: performed deleting, read/unread actions etc. And at some moment Delete/Read/Unread buttons stopped responding. Please, see video and logs below. 

[Status-debug-logs (15).zip](https://github.com/status-im/status-mobile/files/10872373/Status-debug-logs.15.zip)

https://user-images.githubusercontent.com/97245802/222459234-b7d8fce8-5e57-4928-88f0-21a03366a7fc.mp4

**Comment by pavloburykh** [comment]
@ilmotta I have reproduced the above ISSUE 2 before you merged fix of delay reducing. I will install new build and see if the bug is reproducible. But the main problem is that I still not aware about clear steps of reproducing. So I hope you will be able to see some details from logs I attached above.

**Comment by ilmotta** [comment]
> @ilmotta I have reproduced the above ISSUE 2 before you merged fix of delay reducing. I will install new build and see if the bug is reproducible. But the main problem is that I still not aware about clear steps of reproducing. So I hope you will be able to see some details from logs I attached above.

Ouch, I'll try to investigate this problem too. Hopefully we'll be able to reproduce with clear steps.

For the moment I pushed the fix where the chat would not open correctly @pavloburykh.

**Comment by pavloburykh** [comment]
@ilmotta and one more question regarding

> 1. Fixed regression where pressing on a notification would not open the chat.

Currently redirection to the chat works on pressing CR notification that has been accepted by user, thanx for this fix. 

At the same time we have another CR notification which appears after User's A CR has been accepted by another User B. Currently this notification is not tappable, i.e. User is not redirected to chat on pressing this type of notif (see video below). Is this expected behaviour or we should fix and addd redirection on pressing? Maybe you know some details regarding it. 


https://user-images.githubusercontent.com/97245802/222464198-fceb9ac9-6326-4944-a5fb-2c33a85ca936.mp4

**Comment by ilmotta** [comment]
> Currently redirection to the chat works on pressing CR notification that has been accepted by user, thanx for this fix.

:wink: :fireworks:  

> At the same time we have another CR notification which appears after User's A CR has been accepted by another User B. Currently this notification is not tappable, i.e. User is not redirected to chat on pressing this type of notif (see video below). Is this expected behaviour or we should fix and addd redirection on pressing? Maybe you know some details regarding it.

@pavloburykh, the fact that the notification is not tappable very likely means the logic to detect if the notification has been accepted needs to be improved. Currently the code checks only the contact request state, otherwise falls back to a non-tappable incoming contact request. I can fix this issue in a separate PR if you don't mind, because I didn't touch this part of the logic in this PR.

**Comment by pavloburykh** [comment]
> @pavloburykh, the fact that the notification is not tappable very likely means the logic to detect if the notification has been accepted needs to be improved. Currently the code checks only the contact request state, otherwise falls back to a non-tappable incoming contact request. I can fix this issue in a separate PR if you don't mind, because I didn't touch this part of the logic in this PR.

@ilmotta thanx for the explanation. For sure, this improvement is not scope of this PR, I will log it later as separate issue. Just wanted to clarify what behaviour is expected for those types of notifs.

**Comment by pavloburykh** [comment]
> Hi @pavloburykh. Just a couple minutes ago I merged the fix for the issue demonstrated in the video you shared.

@ilmotta Unfortunately I am still reproducing this UI freeze issue https://github.com/status-im/status-mobile/issues/15230  in latest PR build which I suppose should already include the fix https://github.com/status-im/status-mobile/pull/15232 from develop branch.

**Comment by ilmotta** [comment]
> > Hi @pavloburykh. Just a couple minutes ago I merged the fix for the issue demonstrated in the video you shared.
> 
> @ilmotta Unfortunately I am still reproducing this UI freeze issue #15230 in latest PR build which I suppose should already include the fix #15232 from develop branch.

I just made a comment there @pavloburykh, but copying here: do you know if this issue can be reproduced on `develop`? If not then we know the problem is related to this open PR.

**Comment by pavloburykh** [comment]
> Ouch, I'll try to investigate this problem too. Hopefully we'll be able to reproduce with clear steps.

@ilmotta  If you don't mind. I would spend some more time tomorrow and try to figure out the steps. For now I haven't managed to reproduce it in the latest PR build but it doesn't mean that bug is fixed. BTW, do those logs which I attached to the issue contain any valuable details which can help to figure out the rootcause of the problem?

**Comment by ilmotta** [comment]
> > Ouch, I'll try to investigate this problem too. Hopefully we'll be able to reproduce with clear steps.
> 
> @ilmotta If you don't mind. I would spend some more time tomorrow and try to figure out the steps. For now I haven't managed to reproduce it in the latest PR build but it doesn't mean that bug is fixed. 

Sure @pavloburykh, take your time. I think the issue is still there as you said. It's the only one really blocking this PR from advancing, so I really want to fix it, hopefully today or tomorrow.

>BTW, do those logs which I attached to the issue contain any valuable details which can help to figure out the rootcause of the problem?

Unfortunately no, I couldn't figure out anything from the logs. In the video you shared we can see only the swipeable buttons are not interactive, which tells me it's something related to React Native Gesture Handler (a library we use) or how we use it for the swipe buttons. But let's see what we find out :pray:

**Comment by ilmotta** [comment]
@pavloburykh, about the swipe buttons doing nothing, unfortunately I couldn't arrive at the same inconsistent state you shared on the video. I tried with dozens of notifications of all kinds, tried to switch tabs quickly, mark and unmark as read, mark all as read + undo, etc. I also tried on a real device, but nothing unfortunately.

Without being able to reproduce I can't do much, if anything at all. I hope you do find a way to uncover this bug in a deterministic way. If you can't get into that state anymore or just very rarely, what do you think of logging it as a separate issue? Additionally, in case you do reproduce the issue, could you please check if `develop` has the same problem? If you find out `develop` also has the same problem, then I can more easily fix the issue on a separate PR.

In any case, later on, with more users dogfooding the app on a variety of devices we might get clues about the root cause, if the bug is ever reported by somebody.

Cheers!

**Comment by pavloburykh** [comment]
@ilmotta regarding this issue

> ### ISSUE 2 Delete/Read/Unread swipe notification buttons stop responding at some moment in AC

I have reproduced it couple of times in the latest PR builds. But still haven't managed to find out steps. This issue might not be PR related, I haven't checked it in develop as there are not steps in it takes a lot of time in order to reproduce it. I propose not to waste anymore time on it and merge the PR as there are no other issues. 

I will try to investigate this bug later in the develop. Maybe some of other QAs will face it and find out the steps.

**Comment by pavloburykh** [comment]
Thank you for your work @ilmotta

**Comment by ilmotta** [comment]
> @ilmotta regarding this issue
> 
> > ### ISSUE 2 Delete/Read/Unread swipe notification buttons stop responding at some moment in AC
> 
> I have reproduced it couple of times in the latest PR builds. But still haven't managed to find out steps. This issue might not be PR related, I haven't checked it in develop as there are not steps in it takes a lot of time in order to reproduce it. I propose not to waste anymore time on it and merge the PR as there are no other issues.
> 
> I will try to investigate this bug later in the develop. Maybe some of other QAs will face it and find out the steps.

I'm sure you found something nasty. I'm hoping you or the rest of the QA team can pinpoint what could be the cause. I'll keep an eye as well during development.

> Thank you for your work @ilmotta

No, thank you for the lengthy investigation :mechanical_arm:

**Pull request merged**
